exercises/defining_models/exercise.js

```diff
@@ -0,0 +1,259 @@
+    if (process.env.DEBUG) {
+      proc.stdout.pipe(process.stdout, {end: false})
+      proc.stderr.pipe(process.stderr, {end: false})
+    }
+
+    return proc.on('exit', code => {
+      if (code !== 0) {
+        return ready(new Error(`Caught error running postgres. Is docker installed and running? If not, grab it from https://www.docker.com/products/docker-desktop`))
+      }
+
+      getpgport()
+    })
+  }
+
+  function getpgport () {
+    const proc = child_process.spawn('docker', [
+      'port',
+      DOCKER_IMAGE_NAME
+    ])
+    const acc = []
+    proc.stdout.on('data', data => acc.push(data))
+    proc.on('exit', code => {
+      if (code !== 0) {
+        return ready(new Error('could not find pgport'))
+      }
+
+      const output = Buffer.concat(acc).toString()
+      try {
+        pgport = output.split('->').slice(-1)[0].trim().split(':').slice(-1)[0].trim()
+      } catch (err) {
+        return ready(new Error(`"docker port" output in unexpected format: ${output}`))
+      }
+
+      migrate()
+    })
+  }
+
+  async function migrate () {
+    const fixture = fs.readFileSync(path.join(__dirname, 'fixture.sql'), 'utf8')
+
+    process.env.PGURL = `postgres://postgres:postgres@127.0.0.1:${pgport}/postgres`
+    let error = null
+    for (const attempt of [0, 1, 2, 3]) {
+      await new Promise(resolve => setTimeout(resolve, 500 * attempt))
+
+      try {
+        client = new Client({ connectionString: `postgres://postgres:postgres@127.0.0.1:${pgport}/postgres` })
+        await client.connect()
+      } catch (err) {
+        error = err
+      }
+    }
+
+    if (!client) {
+      return ready(error)
+    }
+
+    try {
+      await client.query(fixture)
+    } catch (err) {
+      return ready(err)
+    }
+
+    client.end()
+    checkserver()
+  }
+
+  async function checkserver () {
+
+    function cleanup(err, okay) {
+      if (pgClient) {
+        pgClient.end()
+      }
+      ready(err, okay)
+    }
+
+    const boltzpath = path.join(path.dirname(submission), 'boltzmann.js')
+    var boltzmann = null
+    try {
+      boltzmann = require(boltzpath)
+    } catch (err) {
+      exercise.emit('fail', `Could not load ${boltzpath.replace(process.cwd(), '.')} – caught ${err.stack}`)
+      return cleanup(null, false)
+    }
+
+    var response = null
+    var pgClient = null
+    const wrapped = boltzmann.middleware.test({
+      after: () => {}
+    })(async assert => {
+      pgClient = assert.pgClient
+      response = await assert.request({ url: `/books` })
+    })
+
+    try {
+      await wrapped({})
+    } catch (err) {
+      exercise.emit('fail', `Could not request GET /books – caught ${err.stack}`)
+      return cleanup(null, false)
+    }
+
+    if (response.statusCode !== 200) {
+      exercise.emit('fail', `/books did not respond with 200; got ${response.statusCode}; ${response.payload}`)
+
+      return cleanup(null, false)
+    }
+
+    var books
+    try {
+      books = JSON.parse(response.payload)
```

Comment:
I ended up going with `response.json` which is a few fewer keystrokes, but JSON.parse works as well